# Homework 1

In [1]:
import pandas as pd 
import numpy as np

## Q1. Pandas version
What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.2.3'

**Answer:** 2.2.3

## Getting the data

In [3]:
data = pd.read_csv('laptops.csv')
data.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

## Q2. Records count
How many records are in the dataset?

In [4]:
data.shape[0]

2160

**Answer:** 2160

## Q3. Laptop brands
How many laptop brands are presented in the dataset?

In [5]:
len(data['Brand'].unique())

27

**Answer:** 27

## Q4. Missing values
How many columns in the dataset have missing values?

In [6]:
data.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

**Answer:** 2

## Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

In [7]:
data[data['Brand'] == 'Dell']['Final Price'].max()

np.float64(3936.0)

**Answer:** 3936

## Q6. Median value of Screen
1. Find the median value of `Screen` column in the dataset.
2. Next, calculate the most frequent value of the same `Screen` column.
3. Use `fillna` method to fill the missing values in `Screen` column with the most frequent value from the previous step.
4. Now, calculate the median value of `Screen` once again.

Has it changed?

In [8]:
data['Screen'].median()

np.float64(15.6)

In [9]:
mode = data['Screen'].mode()[0]
mode

np.float64(15.6)

In [10]:
data.fillna({'Screen': mode}, inplace=True);

In [11]:
data['Screen'].median()

np.float64(15.6)

**Answer:** No

## Q7. Sum of weights
1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns `RAM`, `Storage`, `Screen`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.

What's the sum of all the elements of the result?

In [12]:
X_data = data[data['Brand'] == 'Innjoo'][['RAM', 'Storage', 'Screen']]
X_data


RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [13]:
X = X_data.to_numpy()
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [14]:
XTX = np.dot(X.T, X)
XTX_inv = np.linalg.inv(XTX)

In [15]:
y = [1100, 1300, 800, 900, 1000, 1100]
w = np.dot(np.dot(XTX_inv, X.T), y)


In [16]:
w.sum()

np.float64(91.2998806299555)

**Answer:** 91.30